src/main/java/com/limechain/sync/warpsync/state/VerifyJustificationState.java

```diff
@@ -0,0 +1,127 @@
+            // TODO: Throw error if not verified!
+            boolean verified = fragment.getJustification().verify(sync.getAuthoritySet(), sync.getSetId());
+            if (!verified) {
+                throw new Exception("Justification could not be verified.");
+            }
+
+            // Set the latest finalized header and number
+            // TODO: Persist header to DB?
+            sync.setLastFinalizedBlockHash(fragment.getJustification().targetHash);
+            sync.setLastFinalizedBlockNumber(fragment.getJustification().targetBlock);
+
+            try {
+                handleAuthorityChanges(sync, fragment);
+                log.log(Level.INFO, "Verified justification. Block hash is now at #"
+                        + sync.getLastFinalizedBlockNumber() + ": " + sync.getLastFinalizedBlockHash().toString());
+            } catch (Exception error){
+                this.error = error;
+            }
+        } catch (Exception e) {
+            log.log(Level.WARNING, "Error while verifying justification: " + e.getMessage());
+            this.error = e;
+        }
+    }
+
+    public void handleAuthorityChanges(WarpSyncMachine sync, WarpSyncFragment fragment) {
+        // Update authority set and set id
+        AuthoritySetChange authorityChanges;
+        for (HeaderDigest digest : fragment.getHeader().getDigest()) {
+            if (digest.getId() == ConsensusEngine.GRANDPA) {
+                ScaleCodecReader reader = new ScaleCodecReader(digest.getMessage());
+                GrandpaMessageType type = GrandpaMessageType.fromId(reader.readByte());
+
+                switch (type) {
+                    case SCHEDULED_CHANGE:
+                        ScheduledChangeReader authorityChangesReader = new ScheduledChangeReader();
+                        authorityChanges = authorityChangesReader.read(reader);
+
+                        sync.getScheduledAuthorityChanges()
+                                .add(new Pair<>(authorityChanges.getDelay(), authorityChanges.getAuthorities()));
+                        return;
+                    case FORCED_CHANGE:
+                        ForcedChangeReader authorityForcedChangesReader = new ForcedChangeReader();
+                        authorityChanges = authorityForcedChangesReader.read(reader);
+
+                        sync.getScheduledAuthorityChanges()
+                                .add(new Pair<>(authorityChanges.getDelay(), authorityChanges.getAuthorities()));
+                        return;
+                    case ON_DISABLED:
+                        log.log(Level.SEVERE, "On disabled grandpa message not implemented");
+                        return;
+                    case PAUSE:
+                        log.log(Level.SEVERE, "Pause grandpa message not implemented");
+                        return;
+                    case RESUME:
+                        log.log(Level.SEVERE, "Resume grandpa message not implemented");
```

Comment:
I think it's better if we add single quotation marks around the message type, it is a bit confusing at first - "On disabled grandpa..." -> "'On disabled' grandpa..." and so on